File: core/src/main/java/com/lomekwi/cave/task/TaskPool.java
package com.lomekwi.cave.task;

import static com.lomekwi.cave.util.i18n.I18N.i18n;

import com.badlogic.gdx.Gdx;
import com.lomekwi.cave.app.App;
import com.lomekwi.cave.ui.Root;

import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

public class TaskPool implements Iterable<Task>{
    private final Set<Task> tasks = new HashSet<>();
    public void submit(Task task) {
        tasks.add(task);
        App.workerExecutor.submit(() -> {
            try(task) {
                task.run();
            }catch (Exception e){
                // FIXME:ignore
            }
            tasks.remove(task);
            Gdx.app.postRunnable(()-> Root.getInstance().getToastManager().show(i18n("任务")+task.getName()+i18n("已完成")));
        });
    }

    @Override
    public Iterator<Task> iterator() {
        return tasks.iterator();
    }
}

File: core/src/main/scala/com/lomekwi/cave/task/TaskPool.scala
package com.lomekwi.cave.task

import com.lomekwi.cave.util.i18n.I18N.i18n

import com.badlogic.gdx.Gdx
import com.lomekwi.cave.app.App


import scala.util.Using
import java.util

class TaskPool extends java.lang.Iterable[Task] {
  private final val tasks: util.Set[Task] = new util.HashSet[Task]()

  def submit(task: Task): Unit = {
    tasks.add(task)
    App.workerExecutor.execute(() => {
      try {
        Using.resource(task) { t =>
          t.run()
        }
      } catch {
        case e: Exception =>
          // FIXME:ignore
      }
      tasks.remove(task)
      Gdx.app.postRunnable(() => App.root.getToastManager.show(i18n("任务") + task.getName + i18n("已完成")))
    })
  }

  override def iterator(): util.Iterator[Task] = {
    tasks.iterator()
  }
}
